src/tmpfiles/tmpfiles.c

```diff
@@ -1279,21 +1296,29 @@ static int write_one_file(Item *i, const char *path) {
 }
 
 static int create_file(Item *i, const char *path) {
-        _cleanup_close_ int fd = -1;
-        int flags, r = 0;
+        _cleanup_close_ int fd = -1, dir_fd = -1;
         struct stat stbuf, *st = NULL;
+        int flags, r = 0;
+        char *bn;
 
         assert(i);
         assert(path);
         assert(i->type == CREATE_FILE);
 
         /* 'f' operates on regular files exclusively. */
 
+        /* Validate the path and keep the fd on the directory for opening the
+         * file so we're sure that it can't be changed behind our back. */
+        dir_fd = path_open_parent_safe(path);
+        if (dir_fd < 0)
+                return dir_fd;
+
+        bn = basename(path);
```

Comment:
hmm, we should probably be stricter here, and refuse basenames here that aren't proper filenames, i.e. filter out "." and ".." and so on loudly here. i.e. use filename_is_valid() here.

Reply:
This is now done in path_open_parent_safe().